src/data/nat/digits.lean

```diff
@@ -236,10 +238,71 @@ begin
   { simp [of_digits, list.sum_cons, ih], }
 end
 
-/-! ### Inequalities
+/-!
+### Properties
+
+This section contains various lemmas of properties relating to `digits` and `of_digits`.
+-/
+
+lemma digits_eq_nil_iff_eq_zero (b n : ℕ) : digits b n = [] ↔ n = 0 :=
+begin
+  split,
+  { intro h,
+    have : of_digits b (digits b n) = of_digits b [], by rw h,
+    convert this,
+    rw of_digits_digits },
+  { rintro rfl,
+    simp }
+end
+
+lemma digits_ne_nil_iff_ne_zero (b n : ℕ) : digits b n ≠ [] ↔ n ≠ 0 :=
+⟨λ h, h ∘ (digits_eq_nil_iff_eq_zero _ _).mpr,
+ λ h, h ∘ (digits_eq_nil_iff_eq_zero _ _).mp⟩
+
+private lemma digits_last_aux1 {α : Type*} {L M : list α} (h : L = M) (p q) : L.last p = M.last q :=
+by simp [h]
 
-This section contains various lemmas with inequalities relating to `digits` and `of_digits`.
- -/
+private lemma digits_last_aux2 (b n : ℕ) (h : 2 ≤ b) (w : 0 < n) :
+  digits b n = ((n % b) :: digits b (n / b)) :=
+begin
+  rcases b with _|_|b,
+  { finish },
+  { norm_num at h },
+  rcases n with _|n,
+  { norm_num at w },
+  simp,
+end
+
+lemma digits_last (b : ℕ) (h : 2 ≤ b) (m : ℕ) (hm : 0 < m) (p q) :
+  (digits b m).last p = (digits b (m/b)).last q :=
+begin
+  rw digits_last_aux1 (digits_last_aux2 b m h hm),
+  { rw list.last_cons, },
+  exact list.cons_ne_nil _ _,
+end
+
+lemma last_digit_ne_zero (b m : ℕ) (hm : m ≠ 0) (p):
+  (digits b m).last p ≠ 0 :=
```

Comment:
This lemma would be more convenient if you remove the argument `p` and explicitly prove it using `digits_ne_nil_iff_ne_zero.mpr hm`.

Reply:
To be honest that was my first approach to use `digits_ne_nil_iff_ne_zero.mpr hm`. I really struggled to get `nat.strong_induction_on` to work with it, hence I've generalized over it. (see the use of `revert` in the proof).

I can mark this as `private` and add in a lemma with the proof provided, would that be ok? So change this to `private lemma last_digit_ne_zero_aux` and add

```
lemma last_digit_ne_zero {b m : ℕ} (hm : m ≠ 0) :
  (digits b m).last (digits_ne_nil_iff_ne_zero.mpr hm) ≠ 0 :=
last_digit_ne_zero_aux b m hm (digits_ne_nil_iff_ne_zero.mpr hm)
```

Reply:
I've done the above. So marked this one as `private lemma last_digit_ne_zero_aux` and then added another lemma that fills in the proof.

Reply:
That works. Another option was to use as first tactic `generalize : digits_ne_nil_iff_ne_zero.mpr hm = p,`